### PR TITLE
Search for BCC

### PR DESCRIPTION
**Pull request by KillerX**
Proposal for a unified BCC search server.

Please note: This is a proposal and feedback is welcome!

### CODE REVIEW

File: docs/01-search-server.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+1. be fast and scalable (i.e. we must be able to start small and grow with minimal concern for performance issues when scaled up)
+2. have a billing model we can understand and control
+	* Think that "search as you type" can produce *many* queries
+	* Preferably something that does not scale linearly with number of requests
+1. be a mature product that we have reason to believe will be maintained for the next 3-5 years at least
+2. be able to comply with GDPR (EU hosting at a minimum)
+3. be reasonably easy to use for a developer (no need a math degree to understand what's going on)
+
+It should support the following features (loosely ordered by priority):
+
```
Comment (rvanoord) [inline docs/01-search-server.md]: Do we have any requirements or "would likes" regarding analysis of multimedia other than text. E.g. sound or image?

File: docs/01-search-server.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+## Requirements
+
+The system must:
+1. be fast and scalable (i.e. we must be able to start small and grow with minimal concern for performance issues when scaled up)
+2. have a billing model we can understand and control
+	* Think that "search as you type" can produce *many* queries
+	* Preferably something that does not scale linearly with number of requests
+1. be a mature product that we have reason to believe will be maintained for the next 3-5 years at least
+2. be able to comply with GDPR (EU hosting at a minimum)
+3. be reasonably easy to use for a developer (no need a math degree to understand what's going on)
```
Comment (rvanoord) [inline docs/01-search-server.md]: Propose adding some sort of "governance" requirement here. E.g. that access can be granted / revoked ideally based on access groups that are synchronized with our identity directory.

File: docs/01-search-server.md (hunk 1 above)
Comment (KillerX) [reply docs/01-search-server.md]: This is a text search. There is no consideration for that. Personally I think that wanting this in here would explode any system in complexity and required resources. You can for any of the those use other tools to "extract" the information on a textual form (vector embedding is a "textual" form) and then use the search server to do the vector search (f.ex. elastic supports that). But the vectorization itself is a separate concern and should be kept as such IMO.

**Comment by KillerX** [comment]
@fredrikved forgot to tag you before

Also possibly relevant: https://netflixtechblog.com/elasticsearch-indexing-strategy-in-asset-management-platform-amp-99332231e541

If nothing else it shows that Netflix is using it.

File: docs/Current/01-search-server.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	* Unified search across products
+
+### Background
+Currently, different projects use different search solutions. The goal is to standardize on a single, powerful, and flexible solution.
+
+## Alternatives
+
+Main alternatives that were evaluated a bit closer are: 
+
+* Elastic Search,
```
Comment (kkuepper) [inline docs/Current/01-search-server.md:36]: Elasticsearch, one word

File: docs/Current/01-search-server.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
+# Search Engine Standardization for BCC
+
+**Status**: Proposed
+
+## Description
+Evaluate and decide on a standardized search engine solution for BCC projects.
```
Comment (kkuepper) [inline docs/Current/01-search-server.md:6]: When talking about search, maybe we should list the risk that search might change significantly over the next 5 years. Instead of having a Google-like search box it could be replaced with an AI-chatbot.

File: docs/Current/01-search-server/meilisearch.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+Speed is nothing amazing. Don't see benefit over Elastic.
+
+### Issues
+
+* IDs are not autogenerated. Not a big issue just seems odd
+
+
+## Summary
+
+Somewhat meh experience. Don't feel like it adds anything compared to Elastic.
```
Comment (kkuepper) [inline docs/Current/01-search-server/meilisearch.md:70]: Is there no such document for Elasticsearch?

File: docs/Current/01-search-server.md (hunk 4 above)
Comment (kkuepper) [reply docs/Current/01-search-server.md:6]: Probably Search will always be needed. But it could be that 80% of it's current value is served by a chatbot in the future. So there is a risk that our investment into search could be outsized compared to it's future use and the resources better spent on Large Language Models.

File: docs/Current/01-search-server/meilisearch.md (hunk 5 above)
Comment (KillerX) [reply docs/Current/01-search-server/meilisearch.md:70]: Right, forgot to write that one, will add.

File: docs/Current/01-search-server.md (hunk 4 above)
Comment (KillerX) [reply docs/Current/01-search-server.md:6]: **TLDR**: I do not think time spent here is oversized.

Long form: 

I do believe that this will be the case but we need to wait a year before first commercial, controllable, and stable/reliable products like you describe are out. 

Note that I specify controllable. ChatGPT has made big strides there but we would hardly want a chat bot to be able to for example "find all misquoted bible verses", or "find all snippets of text that can be interpreted as violent", "find all speeches where the speaker is angry", etc. So there is quite some work outstanding in that front IMO.

In addition, the "investment" we are doing now is not super huge either. It is to set up a cluster (a few clicks/lines in tf) and integrate with whatever apps (I understand BMM has made progress there). There is no contractual commitment and we can replace the engine if something new appears on the market that we deem better (maybe add?). I also expect a proper "ai search/chat" product will be even easier to integrate later (i.e. not be so finicky on data in, query formulation) thus even quicker to change, so the threshold for change will be even lower.

I do not think that investing the same amount of time into low level LLMs would yeld *any* usable results. I have spent about 25-50% of the time I used for this ADR on testing a LLM and barely got it to run with pre-trained models in that time. Any training setup would require at least an order of magnitude more time. 
TLDR: I do not think time spent here is oversized.

BMM and Bcc Media also obviously need something better than we have now, so we can't wait for that product to emerge.

In general Elastic supports vector search as well, which is the "proto" stage of the chat things so we can experiment with that.

On the other side I think that following along with AI/LLM/etc is important and we should be (and are) testing/evaluating solutions in that space in parallel.

Comment (kkuepper) [reply docs/Current/01-search-server.md:6]: Fully agree
